pkg/operator/cluster/ceph/osd/pod.go

```diff
@@ -61,10 +65,136 @@ func (c *Cluster) makeDaemonSet(selection rookalpha.Selection, config rookalpha.
+	selection rookalpha.Selection, resources v1.ResourceRequirements, config rookalpha.Config) *batch.Job {
+
+	podSpec := c.podTemplateSpec(devices, selection, resources, config, true /* prepare-only */, v1.RestartPolicyOnFailure)
+	podSpec.Spec.NodeSelector = map[string]string{apis.LabelHostname: nodeName}
+
+	return &batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf(prepareAppNameFmt, nodeName),
+			Namespace:       c.Namespace,
+			OwnerReferences: []metav1.OwnerReference{c.ownerRef},
+			Labels: map[string]string{
+				k8sutil.AppAttr:     appName,
+				k8sutil.ClusterAttr: c.Namespace,
+			},
+		},
+		Spec: batch.JobSpec{
+			Template: podSpec,
+		},
+	}
+}
+
+func (c *Cluster) makeOSDDeployment(nodeName string, devices []rookalpha.Device, selection rookalpha.Selection, resources v1.ResourceRequirements, osd OSDInfo) *extensions.Deployment {
+	replicaCount := int32(1)
+	volumeMounts := []v1.VolumeMount{
+		{Name: k8sutil.DataDirVolume, MountPath: k8sutil.DataDir},
+		k8sutil.ConfigOverrideMount(),
+	}
+	volumes := []v1.Volume{k8sutil.ConfigOverrideVolume()}
+	if c.dataDirHostPath != "" {
+		// the user has specified a host path to use for the data dir, use that instead
+		dataDirSource := v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: c.dataDirHostPath}}
+		volumes = append(volumes, v1.Volume{Name: k8sutil.DataDirVolume, VolumeSource: dataDirSource})
+	}
+
+	// by default, don't define any volume config unless it is required
+	if len(devices) > 0 || selection.MetadataDevice != "" {
+		// create volume config for the data dir and /dev so the pod can access devices on the host
+		devVolume := v1.Volume{Name: "devices", VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: "/dev"}}}
+		volumes = append(volumes, devVolume)
+		devMount := v1.VolumeMount{Name: "devices", MountPath: "/dev"}
+		volumeMounts = append(volumeMounts, devMount)
+	}
+
+	if len(volumes) == 0 {
+		logger.Warningf("empty volumes")
+		return nil
+	}
+
+	envVars := []v1.EnvVar{
+		nodeNameEnvVar(),
+		k8sutil.PodIPEnvVar(k8sutil.PrivateIPEnvVar),
+		k8sutil.PodIPEnvVar(k8sutil.PublicIPEnvVar),
+	}
+
+	privileged := true
+	runAsUser := int64(0)
+	readOnlyRootFilesystem := false
+	DNSPolicy := v1.DNSClusterFirst
+	if c.HostNetwork {
+		DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+	journalStr := ""
+	if osd.IsFileStore {
+		journalStr = fmt.Sprintf("--osd-journal=%s", osd.Journal)
+	}
+	return &extensions.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf(osdAppNameFmt, osd.ID),
+			Namespace:       c.Namespace,
+			OwnerReferences: []metav1.OwnerReference{c.ownerRef},
+			Labels: map[string]string{
+				k8sutil.AppAttr:     appName,
+				k8sutil.ClusterAttr: c.Namespace,
+			},
+		},
+		Spec: extensions.DeploymentSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: appName,
+					Labels: map[string]string{
+						k8sutil.AppAttr:     appName,
```

Comment:
I would suggest adding a label here for the OSD ID, to make it easy for someone (or for ceph-mgr) to look up the pod for a particular OSD (unless there's already a way to do that lookup efficiently?)

Reply:
ack, will add a label like this
```yaml
{
  "rook.io/ceph-osd-id": 1
}
```

Reply:
done

Reply:
do we need the `rook.io` prefix? I'd vote for the label `ceph-osd-id` to make it simpler to query with `kubectl`